allium/include/Parser/Parser.h

```diff
@@ -7,43 +7,141 @@
+    ParserResult(std::vector<SyntaxError> errors): TaggedUnion<Optional<T>, std::vector<SyntaxError>>(errors) {}
+
+    friend std::ostream& operator<<(std::ostream& stream, const ParserResult<T>& value) {
+        if (value.errored()) {
+            for (auto error: std::get<std::vector<SyntaxError>>(value.wrapped)) {
+                stream << error;
+            }
+        } else {
+            stream << std::get<Optional<T>>(value.wrapped);
+        }
+
+        return stream;
+    }
+
+    friend bool operator==(const ParserResult<T> &lhs, const ParserResult<T> &rhs) {
+        return lhs.wrapped == rhs.wrapped;
+    }
+
+    friend bool operator!=(const ParserResult<T> &lhs, const ParserResult<T> &rhs) {
+        return lhs.wrapped != rhs.wrapped;
+    }
+
+    // Unwrap the value of the result into the provided location if a value is present or add any errors to the provided list
+    // if errors are present. Return true if value or errors are present and false otherwise.
+    bool unwrapResultInto(T& val, std::vector<SyntaxError>& errorsList) {
+        if (errored()) {
+            std::vector resultErrors = std::get<std::vector<SyntaxError>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped);
+            errorsList.insert(std::end(errorsList), std::begin(resultErrors), std::end(resultErrors));
+            return true;
+        } else {
+            return std::get<Optional<T>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped).unwrapInto(val);
+        }
+    }
+
+    // Unwrap the value of the result into the provided location if a value is present or add any errors to the
+    // provided list if errors are present. Return true if value is present and false otherwise.
+    bool unwrapResultGuard(T& val, std::vector<SyntaxError>& errorsList) {
+        if (errored()) {
+            std::vector resultErrors = std::get<std::vector<SyntaxError>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped);
+            errorsList.insert(std::end(errorsList), std::begin(resultErrors), std::end(resultErrors));
+            return false;
+        } else {
+            return std::get<Optional<T>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped).unwrapGuard(val);
+        }
+    }
+
+    template <typename U>
+    U switchOver(std::function<U(T)> handleValue, std::function<U(void)> handleNone, std::function<U(std::vector<SyntaxError>)> handleError) const {
+        if (errored()) {
+            std::vector resultErrors = std::get<std::vector<SyntaxError>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped);
+            return handleError(resultErrors);
+        } else {
+            return std::get<Optional<T>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped).switchOver(handleValue, handleNone);
+        }
+    }
+
+    bool errored() const {
+        return std::holds_alternative<std::vector<SyntaxError>>(TaggedUnion<Optional<T>, std::vector<SyntaxError>>::wrapped);
```

Comment:
Consider `return wrapped.index == 1`

Reply:
Done in 8afd8258ca80ec1c38534ad56bf51f0a199c58bf. Also figured out the issue I was having with not being able to directly reference `wrapped` from the ParserResult class. See: https://stackoverflow.com/questions/1120833/derived-template-class-access-to-base-class-member-data.